### PR TITLE
Gc colours change

### PR DESCRIPTION
**Pull request by sadiqj**
This PR makes a change to the garbage collector colour scheme in the major collector. It removes the existing gray colour to match the scheme used by the multicore major collector.

This note covers the differences between trunk and multicore&#39;s major GC behaviour with respect to colours and overflow, then it lays out how this PR aims to remove the gray colour to aid merging the multicore collector whilst minimising trunk behaviour changes.

### GC Colours

On trunk we currently have:

- **White** - blocks that we haven&#39;t seen referenced by any live ( **black** ) major heap blocks yet
- **Gray** - blocks that have been referenced by a live ( **black** ) major heap block but whose fields we haven&#39;t yet marked
- **Black** - blocks that have been marked as live and whose fields we have also marked

For the purposes of marking, multicore only has the two states of **UNMARKED** and **MARKED** which correspond to the **White** and **Black** trunk states above. This is because marking is concurrent and idempotent. Additionally marking and sweeping by different domains can overlap which is enabled by the new GC colour scheme. For more details: https://arxiv.org/abs/2004.11663

In terms of how these affect the different runtimes:

On trunk when a block is processed it is marked **Gray** and then added to a list of _gray values_ to be later visited.

On multicore when a block is processed it is **MARKED** and then each field is pushed on to the _mark stack_ to be later visited and marked.

### Different strategies for dealing with overflow

Both of these schemes need a way to deal with the _mark stack_ or _gray values_ overflowing. That is when their sizes grow to become greater than some proportion of the major heap - in both cases 1/32th.

On trunk when _gray values_ overflows it discards some of its entries and marks the heap as _impure_ which later in the marking loop causes the heap to be rescanned for gray blocks so that they can be marked.

On multicore when the _mark stack_ overflows it goes through a process whereby entries are scanned to find the minimum number of pools that will result in a 20% reduction in the size of the stack. These pools are then marked as needing a redarken at a later point in time and their entries removed from the stack.

### Differences in gray values and mark stack

_Gray values_ on trunk and the _mark stack_ on multicore differ in their representation slightly. While _gray values_ on trunk is an array of values that are marked gray on multicore the _mark stack_ is stored as an array of _mark entry_ structs which represent a block currently being scanned, an offset index into the object which will be marked next and the maximum number of fields (_Wosize_) for the block.

### Differences in marking behaviour

A subtle difference between trunk and multicore is their behaviour in marking the object graph.

In trunk this goes as follows:

- A value is taken from the _gray values_
- Each of its fields is iterated
  - The field is marked _gray_ and added to _gray values_

In multicore:

- An entry _e_ is taken from the mark stack
- The offset in the entry _e_ is used to construct a new entry _child_
- The entry _e_&#39;s offset is incremented and it is pushed back on to the mark stack
- _e_ is set to _child_ and we continue from step 2

i.e trunk pushes all fields on to the stack first and then descends whereas multicore descends through each field in turn.

This may lead to different overflow behaviours on the same object graph.

#### Mark stack

In this PR we remove the gray value from trunk and we replace the _gray values_ array with multicore&#39;s _mark stack_. One slight change from multicore is that we no longer store the _end_ (_Wosize)_ in the _mark stack_ &#39;s entry as it seems there is no performance benefit from doing so.

#### Marking behaviour

We retain trunk&#39;s breadth-first marking behaviour.

#### Overflow behaviour

We deviate here from multicore&#39;s existing strategy. All heap chunks are added to a skiplist and the _mark stack_ entries are iterated. For each chunk we build up a range in the chunk that needs to be redarkened at a later time. We do this for all _mark stack_ entries and prune the whole mark stack.

Later when the marking loop exhausts the _mark stack_ we iterate through chunks and redarken blocks with black headers that have any non-black fields. This process will only populate up to a quarter of the mark stack, to prevent subsequent overflows. It is incremental and can be called again when the mark stack empties.

Initially we ported multicore&#39;s _addrmap_ based implementation ([https://github.com/ocaml-multicore/ocaml-multicore/blob/parallel\_minor\_gc/runtime/major\_gc.c#L1385](https://github.com/ocaml-multicore/ocaml-multicore/blob/parallel_minor_gc/runtime/major_gc.c#L1385)) and broke the major heap in to _heap ranges_ as heap chunks can vary in size. This actually didn&#39;t offer a performance advantage over the much simpler skiplist-based approach.

#### Performance

The following are a set of macro benchmarks from the Sandmark suite:

image

Of the macro benchmarks, matrix multiply and the two zarith benchmarks show performance regressions. From profiling all three spend nearly all their time in a few hot loops and next to no time in marking. Re-running on a machine with layout randomization shows no difference in performance between this PR and trunk, which points to a loop layout-related effect.

Below are some existing microbenchmarks in the Sandmark suite that did manage to overflow their mark stack. Overall there seems to be little negative impact on them.

image

#### Overflow algorithm

In theory the simple overflow algorithm should be O(n logc) where n is the number of entries in the mark stack and c is the number of heap chunks. This could be a problem for very large heaps, as the mark stack is a proportion of the heap size though the size of heap chunks also grows rapidly as the heap grows in size.

#### Initial mark stack size

We currently have an initial mark stack size of 2048 mark stack entries. At 16 bytes per entry on 64-bit archs, that is 32k. Should this be a parameter tunable through OCAMLRUNPARAM? An alternative is to set the initial mark stack size much lower but allow it to grow beyond 1/32th of the heap up to a certain size.

### CODE REVIEW

File: runtime/major_gc.c [hunk 1]
```diff
@@ -116,34 +130,145 @@ static unsigned long major_gc_counter = 0;
...
+        Chunk_redarken_end(chunk_addr) = me.block;
+      }
+
+      Chunk_requires_redarken(chunk_addr) = 1;
+
+      if( redarken_first_chunk == NULL || redarken_first_chunk > (char*)chunk_addr ) {
+        redarken_first_chunk = (char*)chunk_addr;
+      }
+    } else {
+      mark_stack[out++] = me;
```
Comment (damiendoligez) [inline runtime/major_gc.c]: This looks strange to me. `me.block` is not in the heap, you will never need to darken it, so you should just drop the entry, right?

**Review by damiendoligez — COMMENTED** [review]
Looks good in principle. I have two questions:

1. Why do you say that the current GC does marking in a breadth-first manner? The gray_vals array is used as a stack.

2. When you prune the mark stack, you empty it completely, but when you reload it, you keep some free space. I would like to understand. Isn't it the case that when you empty the stack, you sometimes need to reload it immediately?

**Comment by sadiqj** [comment]
Thanks for taking a look Damien.
 
>     1. Why do you say that the current GC does marking in a breadth-first manner? The gray_vals array is used as a stack.

You're right - this isn't correct. Both approaches mark in a depth-first manner. The only difference is that trunk pushes all fields onto gray_vals first and then descends down whereas multicore descends down one field at a time (https://github.com/ocaml-multicore/ocaml-multicore/blob/parallel_minor_gc/runtime/major_gc.c#L514). I think, starting from the first or last could mean there are graphs that will overflow the mark stack in one algorithm but not the other.

I will amend the PR comments.

>     2. When you prune the mark stack, you empty it completely, but when you reload it, you keep some free space. I would like to understand. Isn't it the case that when you empty the stack, you sometimes need to reload it immediately?

There's two intentions behind this. One is to avoid a situation where we replenish the mark stack and then immediately need to spill it again. I noticed this happened fairly often on multicore with some of sandmark's benchmarks that overflowed their mark stacks repeatedly. 

The second intention is to try to make some progress. The cost of replenishing the mark stack is fairly low and unless we encounter a unfortunate address earlier on in a chunk we've just redarkened, we shouldn't need to redarken those entries again on a subsequent overflow (because the minimum redarken start will now be higher).

There's probably some other things we could do to make the latter even more likely, such as reversing the order of entries when we replenish the mark stack (so lower addresses at the top of the stack) but have been trying to keep things simple for now.

File: runtime/major_gc.c (hunk 1 above)
Comment (sadiqj) [reply runtime/major_gc.c]: Yes, good spot. There's no point having these on the mark stack. Fixing.

**Pull request closed without merging**